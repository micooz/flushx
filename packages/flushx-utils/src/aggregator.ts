export class Aggregator<K, T> {

  dataMap: Map<K, T[]> = new Map();

  aggregateBy(items: T[], iteratee: Iteratee<K, T>): Map<K, T[]> {
    let keys = [ ...this.dataMap.keys() ];

    // store input items to dataMap by iteratee return value
    for (const item of items) {
      const key: K = iteratee(item);
      if (!key) {
        continue;
      }

      const value = this.dataMap.get(key);

      if (!value) {
        this.dataMap.set(key, [ item ]);
      } else {
        value.push(item);
        // data of this key is changed, remove it from keys
        keys[keys.indexOf(key)] = null;
      }
    }

    // this keys is not changed, should be removed from dataMap
    keys = keys.filter(Boolean);

    // clean not changed keys
    for (const key of keys) {
      this.dataMap.delete(key);
    }

    return this.dataMap;
  }

  dispose(): void {
    this.dataMap.clear();
  }

}

export type Iteratee<K, T> = (item: T) => K;
