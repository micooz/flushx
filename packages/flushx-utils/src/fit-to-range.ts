export function fitToRange(num: number, range: Range): number {
  const { min, max } = range;
  if (num < min) {
    return min;
  }
  if (num > max) {
    return max;
  }
  return num;
}

export interface Range {
  min: number;
  max: number;
}
