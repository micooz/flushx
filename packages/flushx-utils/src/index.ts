export * from './aggregator';
export * from './fit-to-range';
export * from './line-decoder';
export * from './map-to-json';
export * from './memorize';
export * from './split-chunks';
export * from './tail';
export * from './logger';
