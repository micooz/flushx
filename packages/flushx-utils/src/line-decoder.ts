import { EventEmitter } from 'events';

export class LineDecoder extends EventEmitter {

  buf = Buffer.alloc(0);

  static events = {
    LINES: 'lines',
  };

  private readonly encoding: string;

  private readonly delimiter: string;

  constructor(encoding = 'utf8', delimiter = '\n') {
    super();
    this.encoding = encoding;
    this.delimiter = delimiter;
  }

  put(buf: Buffer): void {
    this.buf = Buffer.concat([ this.buf, buf ]);

    const lines: Buffer[] = [];
    // eslint-disable-next-line no-constant-condition
    while (true) {
      const index = this.buf.indexOf(this.delimiter);
      if (index < 0) {
        break;
      }
      lines.push(this.buf.slice(0, index));
      this.buf = this.buf.slice(index + 1);
    }

    if (lines.length > 0) {
      this.emit(LineDecoder.events.LINES, lines.map(line => line.toString(this.encoding)));
    }
  }

}
