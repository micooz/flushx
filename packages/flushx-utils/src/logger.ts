import * as signale from 'signale';

const logLevel = process.env.LOG_LEVEL;

signale.config({
  logLevel,
  displayTimestamp: true,
});

export const Logger = signale;
