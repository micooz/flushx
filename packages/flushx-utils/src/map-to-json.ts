export function mapToJson(map: Map<string, string>): object {
  const obj: { [key: string]: string } = {};
  for (const [ key, value ] of map.entries()) {
    obj[key] = value;
  }
  return obj;
}
