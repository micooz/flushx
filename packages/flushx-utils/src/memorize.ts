/* eslint-disable @typescript-eslint/no-explicit-any */
export type Callback<T> = () => T;

export type MemorizeFn<T> = (callback: Callback<T>, deps: any[]) => T;

export function memorize<T>(): MemorizeFn<T> {
  let previousDeps: any[] = [];
  let cachedValue: T;

  return (callback: Callback<T>, deps: any[]): T => {
    const updateDepsAndCache = (): T => {
      previousDeps = deps;
      return cachedValue = callback();
    };

    if (!cachedValue || previousDeps.length !== deps.length) {
      return updateDepsAndCache();
    }

    // shallow compare each dep
    for (let i = 0; i < deps.length; i += 1) {
      if (deps[i] !== previousDeps[i]) {
        return updateDepsAndCache();
      }
    }

    return cachedValue;
  };
}
