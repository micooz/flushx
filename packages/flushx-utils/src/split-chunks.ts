/**
 * split array into chunks
 * @param arr
 * @param chunkSize
 */
export function splitChunks<T>(arr: T[], chunkSize: number): T[][] {
  if (arr.length < 1) {
    return [];
  }

  if (arr.length <= chunkSize) {
    return [ arr ];
  }

  const chunks: T[][] = [];
  const num = Math.ceil(arr.length / chunkSize);

  for (let i = 0; i < num; i++) {
    chunks.push(arr.slice(i * chunkSize, i * chunkSize + chunkSize));
  }

  return chunks;
}
