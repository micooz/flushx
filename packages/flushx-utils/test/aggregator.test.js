import { Aggregator } from '../src/aggregator';

describe('Aggregator#aggregateBy', () => {
  test('no return key', () => {
    const aggregator = new Aggregator();
    aggregator.aggregateBy([ 1, 2, 3 ], () => undefined);
    expect(aggregator.dataMap).toMatchSnapshot();
  });

  test('normal case', () => {
    const aggregator = new Aggregator();
    aggregator.aggregateBy([ 1, 2, 3 ], () => 1);
    aggregator.aggregateBy([ 1, 2, 3 ], () => 2);
    expect(aggregator.dataMap).toMatchSnapshot();
  });
});

describe('Aggregator#dispose', () => {
  test('dispose', () => {
    const aggregator = new Aggregator();
    aggregator.aggregateBy([ 1, 2, 3 ], () => 1);

    expect(aggregator.dataMap.size).toBe(1);
    aggregator.dispose();

    expect(aggregator.dataMap.size).toBe(0);
  });
});
