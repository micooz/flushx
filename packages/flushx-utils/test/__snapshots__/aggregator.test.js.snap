// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aggregator#aggregateBy no return key 1`] = `Map {}`;

exports[`Aggregator#aggregateBy normal case 1`] = `
Map {
  2 => Array [
    1,
    2,
    3,
  ],
}
`;
